MTKA-1452: Let models override with_front

nickcernis opened the pull request:
## Description

Addresses #457 by adding a “Use Permalink Base” option on the model edit screen to override the with_front setting when registering the model's post type.

https://wpengine.atlassian.net/browse/MTKA-1452

## Details / Context

WordPress registers post types with `'rewrite' => [ 'with_front' => true ]` by default.

This means a site using a custom permalink structure of `/posts/%postname%/` sees their ACM post URLs prefixed by `/posts/`: `/posts/your-acm-model-name/your-post/`.

A user can now untick “Use Permalink Base” if they want `/posts/%postname%/` for regular WP posts, but prefer ACM posts to appear at `/your-acm-model-name/your-post/` (without the `/posts/` prefix).

The PR handles this in a way that does not change behavior for existing models (they will continue to use `with_front` unless a user overrides them by editing the model).

This also does not change the default behavior for new models (new models have empty `with_front`, but custom post type registration logic sets it to true if `with_front` is absent). This is consistent with original behavior and WP defaults.

I chose not to add the field to the "create new model" form, as it's not something a user will need to set or change often, and it reduces the size of the create new model form. (i.e. You can only access the setting when editing a model, not when creating one.) Happy to add it there too if desired, though.

## Testing

Includes e2e tests to check field interactivity and unit tests to check `with_front` applies.

### To test manually

1. Set Permalinks to this custom structure at Settings → Permalinks: `/posts/%postname%/`
2. Create a new model with one field and at least one entry.
3. Run a GraphQL query, being sure to include `uri` as one of the field properties (see screenshots below). The uri in the query response should include the `/posts/` prefix by default.
4. Edit the model and untick the new `“Use Permalink Base”` option.
5. Re-run the query. `/posts/` should no longer appear in the uri.

## Screenshots

<img width="916" alt="Screenshot 2022-03-22 at 18 03 11" src="https://user-images.githubusercontent.com/647669/159535366-e78b6973-9d0d-4e78-82db-14b568994536.png">

<img width="545" alt="Screenshot 2022-03-22 at 18 00 37" src="https://user-images.githubusercontent.com/647669/159534893-802d69a0-ba71-4b24-9914-c9bf465329fc.png">

<img width="765" alt="Screenshot 2022-03-22 at 18 02 24" src="https://user-images.githubusercontent.com/647669/159535292-33f22508-5b0f-431f-9f20-07a2f204b15a.png">

<img width="550" alt="Screenshot 2022-03-22 at 18 03 41" src="https://user-images.githubusercontent.com/647669/159535618-39683f3a-c32f-419d-8286-2f616d2cad85.png">

<img width="921" alt="Screenshot 2022-03-22 at 18 04 12" src="https://user-images.githubusercontent.com/647669/159535629-3ae2f4ce-4e8c-4f9c-96b6-e5a4ef721971.png">

## Documentation Changes

n/a

## Dependant PRs

n/a

nickcernis committed:
feat: set with_front when registering post types

nickcernis committed:
chore: set with_front default when unchecked

nickcernis committed:
feat: Use Permalink Base option on model edit form

nickcernis committed:
test: check with_front is used during registration

nickcernis committed:
test: check with_front is honored

nickcernis committed:
test: check with_front is present and interactive

nickcernis committed:
chore: add changelog entry

nickcernis committed:
chore: try setting permalink structure later

This fails in Circle but passes locally.

nickcernis committed:
test: use WP_UnitTestCase's permalink method

Previous attempt was still failing in Circle but passing locally.

nickcernis committed:
test: try setting permalink structure early

nickcernis committed:
test: try original setUp name

nickcernis committed:
Revert "test: try original setUp name"

This reverts commit 3b1ddd15cc85b36f92647c6c0a6ce498fd1b1c81.

nickcernis committed:
fix: reset wp_rewrite in other test tear_down

Because test_flush_rewrite_rules_called_after_updating_model modifies
the global $wp_rewrite it was affecting other tests that depend on
$wp_rewrite state if it ran before them.

Thanks for all the globals, WP.

nickcernis committed:
fix: set permalink structure override after set_up

nickcernis committed:
chore: tweak field label

mindctrl reviewed (dismissed):
This works great! The only thing I noticed that might confuse someone is when you change this setting, navigate to the Rabbits list table and view a post, it returns a 404 because the permalinks aren't flushed.

nickcernis committed:
chore: re-register content types before flushing

To address @mindctrl's comment at https://github.com/wpengine/atlas-content-modeler/pull/468#pullrequestreview-918850445,
where viewing a post immediately after changing the with_front property
would result in a 404.

nickcernis commented:
@mindctrl Thanks for the review and note! Tracked the rewrite flush thing down to update_registered_content_types(). Should be fixed in https://github.com/wpengine/atlas-content-modeler/pull/468/commits/bf98470c52b5665b7420d002bc4b70fc5ca51c13.

mindctrl reviewed (dismissed):
💥

nickcernis committed:
fix: reset wp_post_types on test tear_down

Required now that update_registered_content_types is
re-registering post types correctly when the option
is updated.

nickcernis commented:
Now ready for final review and merge. Sorry for the review churn.
